File: runtimes/swift-5.9/Sources/Runtime/RuntimeRequest.swift
import Foundation

class RuntimeRequest {
    let bodyRaw: String
    let body: Any
    let headers: [String: String]
    let method: String
    let url: String
    let scheme: String?
    let host: String
    let port: Int?
    let path: String
    let queryString: String?
    let query: [String: String]?

    init(
        bodyRaw: String,
        body: Any,
        headers: [String: String],
        method: String,
        url: String,
        scheme: String?,
        host: String,
        port: Int?,
        path: String,
        queryString: String?,
        query: [String: String]?
    ) {
        self.bodyRaw = bodyRaw
        self.body = body
        self.headers = headers
        self.method = method
        self.url = url
        self.scheme = scheme
        self.host = host
        self.port = port
        self.path = path
        self.queryString = queryString
        self.query = query
    }
}

File: runtimes/kotlin/versions/latest/src/main/kotlin/io/openruntimes/kotlin/RuntimeRequest.kt
package io.openruntimes.kotlin

data class RuntimeRequest(
    val method: String,
    val scheme: String,
    val host: String,
    val port: Int,
    val path: String,
    val query: MutableMap<String, String>,
    val queryString: String,
    val headers: MutableMap<String, String>,
    val bodyBinary: ByteArray,
    val url: String,
) {
    val bodyText: String
        get() = String(bodyBinary)
    val bodyJson: MutableMap<String, Any>
        get() = gson.fromJson(bodyText, MutableMap::class.java) as MutableMap<String, Any>
    val bodyRaw: String
        get() = bodyText
    val body: Any
        get() {
            val contentType = headers.getOrDefault("content-type", "text/plain").lowercase()
            val binaryTypes = arrayOf("application/", "audio/", "font/", "image/", "video/")

            if (contentType.startsWith("application/json")) {
                if (bodyBinary.isEmpty()) {
                    val emptyBody: MutableMap<String, Any> = mutableMapOf()
                    return emptyBody
                } else {
                    return bodyJson
                }
            }

            return when {
                binaryTypes.any { contentType.startsWith(it) } -> bodyBinary
                else -> return bodyText
            }
        }
}
